src/Analyzers/AttributeBinding/MatchingAttributeBindingAnalyzer.cs

```diff
@@ -0,0 +1,79 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.DurableTask.Analyzers.AttributeBinding;
+
+/// <summary>
+/// Expected attribute binding for a given parameter type.
+/// </summary>
+public struct ExpectedBinding
+{
+    /// <summary>
+    /// Gets or sets the expected attribute.
+    /// </summary>
+    public INamedTypeSymbol? Attribute { get; set; }
+
+    /// <summary>
+    /// Gets or sets the expected type.
+    /// </summary>
+    public INamedTypeSymbol? Type { get; set; }
+}
+
+/// <summary>
+/// Analyzer that inspects the parameter type of a method to ensure it matches the expected attribute binding.
+/// It expects one parameter in the DiagnosticRule message template, so the analyzer can report the wrong type.
+/// </summary>
+public abstract class MatchingAttributeBindingAnalyzer : DiagnosticAnalyzer
+{
+    /// <inheritdoc/>
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+        context.RegisterCompilationStartAction(
+            ctx =>
+            {
+                KnownTypeSymbols knownTypeSymbols = new(ctx.Compilation);
```

Comment:
I am not super familiar with source generators APIs, but is there a way to share state between our analyzers? Can then init `KnownTypeSymbols` once and re-use it.

Not high priority, just curious if it is possible.

Reply:
I was reading https://github.com/dotnet/roslyn/issues/6324 and seems there is a way by using the `TryGetValue` methods, available throughout the different analysis contexts: https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.diagnostics.compilationstartanalysiscontext.trygetvalue?view=roslyn-dotnet-4.7.0. I am going to create an issue to investigate it later, that's a good suggestion, although I would love to measure whether the performance impact is worth the state sharing. Thanks!

Issue: https://github.com/microsoft/durabletask-dotnet/issues/302